### PR TITLE
address unsafe issues

### PR DESCRIPTION
**Pull request by edaniels**
go 1.19 is picking up a lot of misuse of unsafe pointers in this library when using the `race` flag. This PR doesn't rely on struct layouts for non-trivial types.

- Addresses https://github.com/golang/go/issues/5253#issuecomment-909495001

### CODE REVIEW

**Comment by gen2brain** [comment]
@edaniels Did you test this, and on which platform? As you can see it doesn't build at all.

**Comment by edaniels** [comment]
Tested on Darwin. Looks like using nested, unnamed structs will be difficult due to auto naming. I'll wrap them in the intermediate C file and then I'll update the PR!

**Comment by edaniels** [comment]
@gen2brain tested on linux and mac now

File: context.go [hunk 1]
```diff
@@ -52,8 +48,24 @@ type ContextConfig struct {
 	Jack                JackContextConfig
 }
 
-func (d *ContextConfig) cptr() *C.ma_context_config {
-	return (*C.ma_context_config)(unsafe.Pointer(d))
+func (d *ContextConfig) toC() (C.ma_context_config, error) {
+	ctxConfig := C.ma_context_config_init()
```
Comment (edaniels) [inline context.go:52]: calling init zeroes out

File: device_config.go [hunk 2]
```diff
@@ -28,11 +30,126 @@ type DeviceConfig struct {
...
+	deviceConfig.wasapi.noHardwareOffloading = C.uchar(d.Wasapi.NoHardwareOffloading)
+
+	deviceConfig.alsa.noMMap = C.uint(d.Alsa.NoMMap)
+	deviceConfig.alsa.noAutoFormat = C.uint(d.Alsa.NoAutoFormat)
+	deviceConfig.alsa.noAutoChannels = C.uint(d.Alsa.NoAutoChannels)
+	deviceConfig.alsa.noAutoResample = C.uint(d.Alsa.NoAutoResample)
+
+	var releasers []func()
+	if d.Pulse.StreamNameCapture != "" {
+		streamNameCapturePtr := C.CString(d.Pulse.StreamNameCapture)
```
Comment (edaniels) [inline device_config.go:134]: this is now actually able to be used; before they were just *int8s

File: device_config.go [hunk 3]
```diff
@@ -57,14 +173,14 @@ type WasapiDeviceConfig struct {
 type AlsaDeviceConfig struct {
 	NoMMap         uint32
 	NoAutoFormat   uint32
-	NoAutoChannles uint32
+	NoAutoChannels uint32
```
Comment (edaniels) [inline device_config.go:176]: typo

**Comment by edaniels** [comment]
Can't run the workflow here but it's fine over on my fork: https://github.com/edaniels/malgo/runs/7846358123?check_suite_focus=true

**Comment by edaniels** [comment]
Any chance of getting this in this week?

**Comment by gen2brain** [comment]
@edaniels Sorry, I need to test first, the similar PR you referenced worked only on Windows as I remember. I will try to do that this week.

**Comment by gen2brain** [comment]
Merged, thanks!

**Pull request merged**